prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1318 @@
+`send_id` in CTS packet. An endpoint will have multiple send operations at the same time, thus
+when processing a CTS packet from a receive, it needs a way to locate the send operation the
+CTS packet is referring to.
+
+Admittedly, the introduction of `send_id` is not absolute necessary, because receiver could have
+included `msg_id` in CTS header, and sender should be able to locate the send operation using
+the combination of receiver's address and message ID. However, that approach would require
+the sending endpoint set up a map between (address + `msg_id`) and send operation, and look up the map every time
+it received a CTS packet. We considered that approach too burdensome for an endpoint to implmement
+and decided to introduce a 4 byte `send_id` in LONGCTS_RTM header to eliminate the cost.
+
+Another note about `send_id` is that it can be reused between messages. Because `send_id` is used to
+distinguish on-the-fly TX operations, so a send operation may have the same `send_id` as a previous
+one that has already finished.
+
+The field `send_id` was named `tx_id` when the protocol was initially introduced. It is renamed
+because the new name is clearer.
+
+The field `credit_request` is how many DATA packets the sender wish to receive from the receiver,
+the receiver will try to honor the request, but is not obligated to. However, receiver must allow
+the sender to send at least 1 DATA packet back, to keep the communication moving forward.
+
+Besides the LONGCTS_RTM packet, there are two other packet types used by the long-CTS message protocol:
+CTS and DATA.
+
+The binary format of a CTS packet is listed in table 3.6:
+
+Table: 3.6 the binary format a CTS packet
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `pad`            | 4 | array   | `uint8_t`  | used for 8 byte alignment |
+| `send_id`        | 4 | integer | `uint32_t` | send id from LONGCTS_RTM |
+| `recv_id`        | 4 | integer | `uint32_t` | receive id to be used in DATA packet |
+| `recv_length`    | 8 | integer | `uint64_t` | number of bytes the receiver is ready to receive |
+| `connid`         | 4 | integer | `uint32_t` | optional sender connection id (see chapter 4.4 for more detail) |
+
+The 3 new fields in the header are `pad`, `recv_id` and `recv_length`.
+
+The field `pad` is 4 byte padding to make the variables align on 8 bytes boundary.
+
+The field `recv_id` is similar to `send_id` introduced earlier, but for an on-going receive operation.
+Sender should include `recv_id` in the DATA packet.
+
+The field `recv_length` is the number of bytes receiver is ready to receive for this operation,
+it must be > 0 to make the communication going forward.
```

Comment:
"it must be >0 otherwise the receiver should wait to post the CTS packet" might be clearer here